src/charm.py

```diff
@@ -67,17 +70,28 @@ def _on_config_changed(self, event) -> None:
         self.configure_workload_container(event)
 
     def _report_visits_by_ip_action(self, event: ActionEvent) -> None:
-        """Report requests to nginx by IP and report action result."""
+        """Handle the report-visits-by-ip action."""
+        results = self._report_visits_by_ip(event)
+        event.set_results({"ips": str(results)})
```

Comment:
Would it be worth formatting this report for the user a bit? Like maybe converting it to a nice table or something like that?

Reply:
definitely. Printing a stringified list of tuples is not the best option, but it seems juju actions has some problems printing other data types (see previous Arturo's comments and my answers with pastebin). I'll file a bug regarding this, but in the meanwhile I'll leave it as it is, does that make sense?

Reply:
We discussed this as an option yesterday: https://pypi.org/project/tabulate/